app/uk/gov/hmrc/helptosavefrontend/repo/SessionStore.scala

```diff
@@ -0,0 +1,139 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.helptosavefrontend.repo
+
+import cats.data.{EitherT, OptionT}
+import cats.instances.either._
+import cats.instances.future._
+import cats.syntax.either._
+import com.google.inject.{ImplementedBy, Inject, Singleton}
+import play.api.libs.json.{JsValue, Json, Reads, Writes}
+import play.modules.reactivemongo.ReactiveMongoComponent
+import uk.gov.hmrc.cache.model.Id
+import uk.gov.hmrc.cache.repository.CacheMongoRepository
+import uk.gov.hmrc.helptosavefrontend.config.FrontendAppConfig
+import uk.gov.hmrc.helptosavefrontend.metrics.Metrics
+import uk.gov.hmrc.helptosavefrontend.models.HTSSession
+import uk.gov.hmrc.helptosavefrontend.util.{Result, toFuture}
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@ImplementedBy(classOf[SessionStoreImpl])
+trait SessionStore {
+
+  def get(implicit reads: Reads[HTSSession], hc: HeaderCarrier): Result[Option[HTSSession]]
+
+  def store(body: HTSSession)(implicit writes: Writes[HTSSession], hc: HeaderCarrier): Result[Unit]
+}
+
+@Singleton
+class SessionStoreImpl @Inject() (mongo: ReactiveMongoComponent, metrics: Metrics)(implicit appConfig: FrontendAppConfig, ec: ExecutionContext) extends SessionStore {
+
+  private val expireAfterSeconds = appConfig.mongoSessionExpireAfter.toSeconds
+
+  private lazy val cacheRepository = new CacheMongoRepository("sessions", expireAfterSeconds)(mongo.mongoConnector.db, ec)
+
+  private type EitherStringOr[A] = Either[String, A]
+
+  override def get(implicit reads: Reads[HTSSession], hc: HeaderCarrier): Result[Option[HTSSession]] = {
+
+    val timerContext = metrics.sessionReadTimer.time()
+
+    EitherT(hc.sessionId.map(_.value) match {
+      case Some(sessionId) ⇒
+        doFindById(Id(sessionId)).map { maybeCache ⇒
+          val response: OptionT[EitherStringOr, HTSSession] = for {
+            cache ← OptionT.fromOption[EitherStringOr](maybeCache)
+            data ← OptionT.fromOption[EitherStringOr](cache.data)
+            result ← OptionT.liftF[EitherStringOr, HTSSession](
+              (data \ "htsSession").validate[HTSSession].asEither.leftMap(e ⇒ s"Could not parse session data from mongo: ${e.mkString("; ")}"))
+          } yield result
+
+          val _ = timerContext.stop()
+
+          response.value
+
+        }.recover {
+          case e ⇒
+            val _ = timerContext.stop()
+            metrics.sessionReadErrorCounter.inc()
+            Left(e.getMessage)
+        }
+
+      case None ⇒
+        // is this correct to return None if no session found in the HeaderCarrier ? or should this be an Left(error) ?
```

Comment:
realistically we shouldn't expect to get to this stage since we shouldn't be looking at session data until we're sure that someone is logged in. After we know they're logged in they should have a session ID. Because of this we should probably be returning an error here since this would be an unexpected event we'd want to investigate